aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/kinesisvideo/http/HostnameVerifyingX509ExtendedTrustManager.java

```diff
@@ -0,0 +1,234 @@
+package com.amazonaws.kinesisvideo.http;
+
+import com.amazonaws.kinesisvideo.common.logging.Log;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509ExtendedTrustManager;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+
+/*
+     Adapted from (Apache 2.0 License):
+
+     https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/common/
+     ZKTrustManager.java
+
+     A custom TrustManager that supports hostname verification via org.apache.http.conn.ssl.DefaultHostnameVerifier.
+     *
+     * We attempt to perform verification using just the IP address first and if that fails will attempt to perform a
+     * reverse DNS lookup and verify using the hostname.
+*/
+
+public class HostnameVerifyingX509ExtendedTrustManager extends X509ExtendedTrustManager {
+
+    private static final HostnameVerifier DEFAULT_HOSTNAME_VERIFIER = HostnameVerifier.INSTANCE;
+    private Log log = new Log(Log.SYSTEM_OUT);
+    private final boolean clientSideHostnameVerificationEnabled;
+
+    private final X509ExtendedTrustManager x509ExtendedTrustManager;
+
+    public HostnameVerifyingX509ExtendedTrustManager(final boolean clientSideHostnameVerificationEnabled) {
+        this.x509ExtendedTrustManager = getX509ExtendedTrustManager();
+        this.clientSideHostnameVerificationEnabled = clientSideHostnameVerificationEnabled;
+    }
+
+    private X509ExtendedTrustManager getX509ExtendedTrustManager() {
+        TrustManagerFactory factory;
+        try {
+            factory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            factory.init((KeyStore) null);
+        } catch (NoSuchAlgorithmException nsae) {
+            log.error("Unable to initialize default TrustManagerFactory, using no-op X509ExtendedTrustManager", nsae);
```

Comment:
Do you know what scenarios would result in the `NoSuchAlgorithmException` or the `KeyStoreException`?   It looks like you essentially disabling hostname verification if either of these excecptions occur, by using a no-op instance, is that right?

Reply:
You're right, we should instead bail if either of those two exceptions are thrown, basically the `NoSuchAlgorithmException` happens if no registered security provider supports a `TrustManagerFactorySpi` implementation of the specified algorithms (https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/TrustManagerFactory.html#getInstance-java.lang.String-).  And the `KeyStoreException` happens when the init fails for some reason (https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/TrustManagerFactory.html#init-java.security.KeyStore-)

Perhaps in this case we may want to call `Security.getProviders()` and relay that in the error message before bailing so we can track down why the failure happened as well as the string `TrustManagerFactory.getDefaultAlgorithm()`

Reply:
I agree, bailing is a better approach than returning a no op instance.